### PR TITLE
[FLINK-27483]Add session job config field

### PR DESCRIPTION
**Pull request by FuyaoLi2017**
Add session job configuration field. Allow user to add custom HTTP header if they use HTTP based Jar fetcher for session jobs.

### CODE REVIEW

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorSessionJobConfigOptions.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package org.apache.flink.kubernetes.operator.config;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+/** List supported session job specific configurations */
+public enum KubernetesOperatorSessionJobConfigOptions {
+
+    SESSION_JOB_HTTP_JAR_HEADERS("kubernetes.operator.user.artifacts.http.header");
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorSessionJobConfigOptions.java:10]: I think this all can be simply part of the `KuberenetesOperatorConfigOptions` no need for a new class

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/crd/spec/FlinkSessionJobSpec.java [hunk 2]
```diff
@@ -38,4 +40,7 @@ public class FlinkSessionJobSpec extends AbstractFlinkSpec {
 
     /** The name of the target session cluster deployment. */
     private String deploymentName;
+
+    /** Session job specific configuration */
+    private Map<String, String> sessionJobFlinkConfiguration;
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/crd/spec/FlinkSessionJobSpec.java:45]: Instead of specificng a new config here, lets move the `flinkConfiguration` field of the `FlinkDeploymentSpec` up to `AbstractFlinkSpec` that way it will be inherited

**Review by gyfora — CHANGES_REQUESTED** [review]
Thank you! Added some initial minor comments

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorSessionJobConfigOptions.java (hunk 1 above)
Comment (FuyaoLi2017) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorSessionJobConfigOptions.java:10]: Hi @gyfora, I was thinking if we put this configuration at the `KuberenetesOperatorConfigOptions`, Does that imply you intend to put this header config field as a candidate in `flink-conf.yaml`? In that case, we are assuming the headers are the same for all session jobs within the cluster? I think maybe we can allow users to config different headers for different session jobs within one session cluster.

Comment (FuyaoLi2017) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorSessionJobConfigOptions.java:10]: Hi @gyfora If not, you also intend to keep this new `flinkConfiguration` as job specific. How do we control the behavior in the session cluster since it could have multiple different jobs?
Taking a step back, we can assume the headers are all the same for the same cluster, that will make things simplier, but might NOT be very flexible.
cc @wangyang0918 
See Yang's comment.
https://issues.apache.org/jira/browse/FLINK-27483?focusedCommentId=17532137&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17532137

Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorSessionJobConfigOptions.java:10]: I am not suggesting to change the logic, just the location of the constant you defined to not create unnecessary new classes and spread the options too much :)

The operator should always use the config defined in the resource (session job/deployment) no difference. We usually apply this logic on top of the default config defined in the operator.

Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorSessionJobConfigOptions.java:10]: But I would allow users to put the header conf in the default Flink conf yaml to have some default value applied in case the session job doesn’t want to ovverride it

Comment (wangyang0918) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorSessionJobConfigOptions.java:10]: We just need to document clearly which configuration options could take effect in `FlinkSessionJob.spec. flinkConfiguration`.

**Comment by FuyaoLi2017** [comment]
Hi @gyfora , I noticed this error.
```
Error:  Failed to execute goal org.apache.rat:apache-rat-plugin:0.12:check (default) on project flink-kubernetes-operator-parent: Too many files with unapproved license: 1 See RAT report in: /home/runner/work/flink-kubernetes-operator/flink-kubernetes-operator/target/rat.txt -> [Help 1]
```
I don't know how to fix this. I can build the project successfully on my local laptop.

**Comment by FuyaoLi2017** [comment]
In the CI pipeline, it shows some unit test failures. But I can pass them in my local. I guess it might be some cascading error caused by documentation? I didn't rebuild the docs eariler since I am not familiar with the procedure.
```
mvn clean install -DskipTests -Pgenerate-docs
```
I ran the command above, generated some docs and pushed again...

**Comment by gyfora** [comment]
Thank you @FuyaoLi2017 I will try to review this latest on monday as I will be travelling during the weekend.

@Aitozi if you have some spare capacity I would appreciate a review but we can also do this on monday :)

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/crd/spec/AbstractFlinkSpec.java [hunk 3]
```diff
@@ -40,4 +42,7 @@ public abstract class AbstractFlinkSpec {
      * restart, change the number to anything other than the current value.
      */
     private Long restartNonce;
+
+    /** Flink configuration overrides for the Flink deployment. */
```
Comment (Aitozi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/crd/spec/AbstractFlinkSpec.java:46]: nit: the comment should be update: overrides for the Flink deployment or session job.

File: docs/content/docs/operations/configuration.md [hunk 4]
```diff
@@ -53,3 +53,11 @@ To learn more about metrics and logging configuration please refer to the dedica
 ## Operator Configuration Reference
 
 {{< generated/kubernetes_operator_config_configuration >}}
+
+## Job Specific Configuration Reference
+Job specific configuration can be configured under `spec.job.flinkConfiguration` and it will override flink configurations defined in `flink-conf.yaml`.
+
+- For application clusters, `spec.job.flinkConfiguration` will be located in `FlinkDeployment` CustomResource.
```
Comment (Aitozi) [inline docs/content/docs/operations/configuration.md:60]: ditto

File: docs/content/docs/operations/configuration.md [hunk 5]
```diff
@@ -53,3 +53,11 @@ To learn more about metrics and logging configuration please refer to the dedica
...
+- For session clusters, configuring `spec.job.flinkConfiguration` in parent `FlinkDeployment` will be applied to all session jobs within the session cluster.
+You can also configure `spec.job.flinkConfiguration` in `FlinkSessionJob` CustomResource for a specific session job.
```
Comment (Aitozi) [inline docs/content/docs/operations/configuration.md:62]: A newline here

File: docs/layouts/shortcodes/generated/kubernetes_operator_config_configuration.html [hunk 6]
```diff
@@ -110,5 +110,11 @@
             <td>String</td>
             <td>The base dir to put the session job artifacts.</td>
         </tr>
+        <tr>
+            <td><h5>kubernetes.operator.user.artifacts.http.header</h5></td>
+            <td style="word-wrap: break-word;">(none)</td>
+            <td>Map</td>
+            <td>Custom HTTP header for a Flink job. If configured in cluster level, headers will be applied to all jobs within the cluster. This field can also be configured under spec.job.flinkConfiguration for a specific session job within a session cluster. If configured at session job level, it will override the cluster level configuration. Expected format: headerKey1:headerValue1,headerKey2:headerValue2.</td>
```
Comment (Aitozi) [inline docs/layouts/shortcodes/generated/kubernetes_operator_config_configuration.html:117]: ditto

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/artifact/HttpArtifactFetcher.java [hunk 7]
```diff
@@ -32,12 +37,24 @@ public class HttpArtifactFetcher implements ArtifactFetcher {
...
 
     @Override
-    public File fetch(String uri, File targetDir) throws Exception {
+    public File fetch(String uri, Configuration flinkConfiguration, File targetDir)
+            throws Exception {
         var start = System.currentTimeMillis();
         URL url = new URL(uri);
+        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+
+        Map<String, String> clusterLevelHeader =
```
Comment (Aitozi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/artifact/HttpArtifactFetcher.java:46]: we do not know this header is cluster level or job level here, what about naming it `headers`  directly ?

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java [hunk 8]
```diff
@@ -83,6 +84,13 @@ public void reconcile(FlinkSessionJob flinkSessionJob, Context context) throws E
 
         Configuration deployedConfig = configManager.getObserveConfig(flinkDepOptional.get());
 
+        // merge session job specific config
+        Map<String, String> sessionJobFlinkConfiguration =
```
Comment (Aitozi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java:88]: nit: maybe we could move this to the `FlinkConfigManager`

File: docs/content/docs/operations/configuration.md [hunk 9]
```diff
@@ -53,3 +53,11 @@ To learn more about metrics and logging configuration please refer to the dedica
...
 
 {{< generated/kubernetes_operator_config_configuration >}}
+
+## Job Specific Configuration Reference
+Job specific configuration can be configured under `spec.job.flinkConfiguration` and it will override flink configurations defined in `flink-conf.yaml`.
+
+- For application clusters, `spec.job.flinkConfiguration` will be located in `FlinkDeployment` CustomResource.
+- For session clusters, configuring `spec.job.flinkConfiguration` in parent `FlinkDeployment` will be applied to all session jobs within the session cluster.
+You can also configure `spec.job.flinkConfiguration` in `FlinkSessionJob` CustomResource for a specific session job. 
+The session job level configuration will override the parent session cluster's Flink configuration.
```
Comment (Aitozi) [inline docs/content/docs/operations/configuration.md:63]: I think we should let user know that the cluster level's flink config will not overridden by the session job's flinkConfiguration actually (I mean the session cluster's config). Personally, I think what sessionjob`spec.flinkConfiguration` defined, is not the **flink** Configuration, just some custom configs.

**Review by Aitozi — COMMENTED** [review]
Thanks @FuyaoLi2017  for your work, I left some minor comments

File: docs/content/docs/operations/configuration.md [hunk 10]
```diff
@@ -53,3 +53,11 @@ To learn more about metrics and logging configuration please refer to the dedica
 ## Operator Configuration Reference
 
 {{< generated/kubernetes_operator_config_configuration >}}
+
+## Job Specific Configuration Reference
+Job specific configuration can be configured under `spec.job.flinkConfiguration` and it will override flink configurations defined in `flink-conf.yaml`.
```
[the remark this answers is not in the document]
Comment (FuyaoLi2017) [reply docs/content/docs/operations/configuration.md:58]: I think this should correct. If you take a look at the CRD, this property will be sitting in `spec.job.flinkConfiguration` from the root of the CRD.

File: docs/layouts/shortcodes/generated/kubernetes_operator_config_configuration.html (hunk 6 above)
Comment (FuyaoLi2017) [reply docs/layouts/shortcodes/generated/kubernetes_operator_config_configuration.html:117]: This is a generated html page. I guess we don't need to touch it.

File: docs/content/docs/operations/configuration.md (hunk 9 above)
Comment (FuyaoLi2017) [reply docs/content/docs/operations/configuration.md]: Added related docs and validator code changes.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java (hunk 8 above)
Comment (FuyaoLi2017) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/FlinkSessionJobReconciler.java]: I created a `getSessionJobObserveConfig` method in `FlinkConfigManager` and replace this one and a few other places.

File: docs/content/docs/operations/configuration.md (hunk 5 above)
Comment (FuyaoLi2017) [reply docs/content/docs/operations/configuration.md]: I added a new line at the end.

**Comment by FuyaoLi2017** [comment]
Hi @Aitozi @gyfora , thanks for reviewing this. I have updated the code based on your suggestions. 

1. Update typos for `spec.job.flinkConfiguration`.
2. Update documentation about the session job flink configuration and provided allowed configurations (only the header config at the moment).
3. Add validator logic in `DefaultValidator`. Also refactored existing `validateFlinkConfig` method name to `validateFlinkDeploymentConfig` to avoid confusion.
4. Move configuration merge logic into `FlinkConfigManager`. Add a `getSessionJobObserveConfig` method, refactor a few places to call this method instead of `getObserveConfig`.

There were some merge conflicts and rebased the main branch. I ran into some errors and introduced some unexpected changes, as a result, I force pushed the code to make it clean. Sorry for making the commit history kind of messy..

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java [hunk 11]
```diff
@@ -141,4 +142,15 @@ public class KubernetesOperatorConfigOptions {
                     .noDefaultValue()
                     .withDescription(
                             "Whether to enable recovery of missing/deleted jobmanager deployments. False by default for Flink 1.14, true for newer Flink version.");
+
+    public static final ConfigOption<Map<String, String>> JAR_ARTIFACT_HTTP_HEADER =
+            ConfigOptions.key("kubernetes.operator.user.artifacts.http.header")
+                    .mapType()
+                    .noDefaultValue()
+                    .withDescription(
+                            "Custom HTTP header for a Flink job. If configured in cluster level, headers will be applied to all jobs within"
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:151]: I think the description should detail what the config is good for. Does it affect artifact downloading?

It doesn't need to explain how configurations work, we have general docs for that. So I would completely remove this part:
```
If configured in cluster level, headers will be applied to all jobs within the cluster. This field can also be configured under spec.job.flinkConfiguration for a specific session job within a session cluster. If configured at session job level, it will override the cluster level configuration.
```

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java [hunk 12]
```diff
@@ -133,6 +135,19 @@ public Configuration getObserveConfig(FlinkDeployment deployment) {
         return getConfig(deployment.getMetadata(), ReconciliationUtils.getDeployedSpec(deployment));
     }
 
+    public Configuration getSessionJobObserveConfig(
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:138]: I think it's a bit confusing to call it `getSessionJobObserveConfig`. I would prefer to call it simply `getSessionJobConfig`

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java [hunk 13]
```diff
@@ -133,6 +135,19 @@ public Configuration getObserveConfig(FlinkDeployment deployment) {
...
+            FlinkDeployment deployment, FlinkSessionJob flinkSessionJob) {
+        Configuration sessionJobConfig = getObserveConfig(deployment);
+
+        // merge session job specific config
+        Map<String, String> sessionJobFlinkConfiguration =
+                flinkSessionJob.getSpec().getFlinkConfiguration();
+        if (sessionJobFlinkConfiguration != null && !sessionJobFlinkConfiguration.isEmpty()) {
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:145]: The `!sessionJobFlinkConfiguration.isEmpty()` check is not necessary

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/validation/DefaultValidator.java [hunk 14]
```diff
@@ -404,4 +409,21 @@ private Optional<String> validateServiceAccount(String serviceAccount) {
...
     }
+
+    private Optional<String> validateFlinkSessionJobConfig(
+            Map<String, String> flinkSessionJobConfig) {
+        if (flinkSessionJobConfig == null) {
+            return Optional.empty();
+        }
+
+        for (String key : flinkSessionJobConfig.keySet()) {
+            if (!ALLOWED_FLINK_SESSION_JOB_CONF_KEYS.contains(key)) {
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/validation/DefaultValidator.java:420]: Could we please add a simple test for this into the validator test?

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/artifact/FileSystemBasedArtifactFetcher.java [hunk 15]
```diff
@@ -33,7 +34,8 @@ public class FileSystemBasedArtifactFetcher implements ArtifactFetcher {
             new FileSystemBasedArtifactFetcher();
 
     @Override
-    public File fetch(String uri, File targetDir) throws Exception {
+    public File fetch(String uri, Configuration flinkConfiguration, File targetDir)
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/artifact/FileSystemBasedArtifactFetcher.java:37]: Why did you add the config field here? Is it used?

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/artifact/HttpArtifactFetcher.java [hunk 16]
```diff
@@ -32,12 +37,26 @@ public class HttpArtifactFetcher implements ArtifactFetcher {
...
         var start = System.currentTimeMillis();
         URL url = new URL(uri);
+        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+
+        // merged session job level header and cluster level header, session job level header take
+        // precedence.
+        Map<String, String> headers =
+                flinkConfiguration.get(KubernetesOperatorConfigOptions.JAR_ARTIFACT_HTTP_HEADER);
+
+        if (headers != null && headers.size() > 0) {
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/artifact/HttpArtifactFetcher.java:51]: size check not necessary

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java [hunk 17]
```diff
@@ -41,6 +42,7 @@ public class FlinkOperatorConfiguration {
     Duration flinkCancelJobTimeout;
     Duration flinkShutdownClusterTimeout;
     String artifactsBaseDir;
+    Map<String, String> artifactHttpHeader;
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java:45]: We shouldn't add this extra field here, it's not an operator configuration and it's also not used anywhere

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java [hunk 18]
```diff
@@ -133,6 +135,19 @@ public Configuration getObserveConfig(FlinkDeployment deployment) {
...
+            sessionJobFlinkConfiguration.forEach(sessionJobConfig::setString);
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:146]: I am wondering for this particular case, would it make sense to merge the HTTP headers instead of overwriting them if the base config also defined it?

**Review by gyfora — CHANGES_REQUESTED** [review]

**Comment by gyfora** [comment]
@FuyaoLi2017 no worries, I have went through the code and added comments. Mostly minor stuff :)

No need to worry about force pushing, this is your branch, it's necessary to force push PR branches to keep them clean.
I also generally squash them together into 1-2  commits

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/artifact/FileSystemBasedArtifactFetcher.java (hunk 15 above)
Comment (FuyaoLi2017) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/artifact/FileSystemBasedArtifactFetcher.java:37]: No, it is not used. I need to change this line since this class overrides the `artifactFetcher` interface. In the future, we might need this configuration? Not sure. Anyways, it doesn't hurt.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java (hunk 18 above)
Comment (FuyaoLi2017) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:146]: I guess the logic would be more straight forward if this config in sessionJob directly overrides the parent config. 

If we do the merging work. Then a user will need to check two CRs instead of one to determine what are the actual headers that is applied during jar fetching during debugging. It could a little bit confusing.

File: docs/content/docs/operations/configuration.md [hunk 19]
```diff
@@ -53,3 +53,12 @@ To learn more about metrics and logging configuration please refer to the dedica
...
 {{< generated/kubernetes_operator_config_configuration >}}
+
+## Job Specific Configuration Reference
+Job specific configuration can be configured under `spec.flinkConfiguration` and it will override flink configurations defined in `flink-conf.yaml`.
+
+- For application clusters, `spec.flinkConfiguration` will be located in `FlinkDeployment` CustomResource.
+- For session clusters, configuring `spec.flinkConfiguration` in parent `FlinkDeployment` will be applied to all session jobs within the session cluster.
+  - You can configure some additional job specific supplemental configuration through `spec.flinkConfiguration` in `FlinkSessionJob` CustomResource. 
+  Those session job level configurations will override the parent session cluster's Flink configuration. Please note only the following configurations are considered to be valid configurations.
+    - `kubernetes.operator.user.artifacts.http.header`
```
Comment (Aitozi) [inline docs/content/docs/operations/configuration.md:64]: I think it's much clear now 👍🏻. 
Some other thoughts come to my mind that, Maybe we could extend the `ConfigOption` to mark the config belong to which component (flink/operator). It's out of this PR, we could do it later if necessary.

**Review by Aitozi — APPROVED** [review]
LGTM, Nice work 👍🏻

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java [hunk 20]
```diff
@@ -141,4 +142,11 @@ public class KubernetesOperatorConfigOptions {
                     .noDefaultValue()
                     .withDescription(
                             "Whether to enable recovery of missing/deleted jobmanager deployments. False by default for Flink 1.14, true for newer Flink version.");
+
+    public static final ConfigOption<Map<String, String>> JAR_ARTIFACT_HTTP_HEADER =
+            ConfigOptions.key("kubernetes.operator.user.artifacts.http.header")
+                    .mapType()
+                    .noDefaultValue()
+                    .withDescription(
+                            "Custom HTTP header for a Flink job. Expected format: headerKey1:headerValue1,headerKey2:headerValue2.");
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:151]: Could we change this to: 
```
Custom HTTP header for HttpArtifactFetcher. The header will be applied when getting the session job artifacts.  Expected format: headerKey1:headerValue1,headerKey2:headerValue2.
```

**Review by gyfora — APPROVED** [review]
I had a last minor comment about the config description :) 

After that it's good to merge!

Thank you for your work.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java (hunk 20 above)
Comment (FuyaoLi2017) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java]: Sure. I just pushed the change.
